vm/systemSmartContracts/esdt.go

```diff
@@ -224,94 +230,366 @@ func (e *esdt) issueToken(owner []byte, arguments [][]byte) error {
-	//TODO: implement me
+func (e *esdt) wipe(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+	if len(args.Arguments) < 2 {
+		e.eei.AddReturnMessage("invalid number of arguments, wanted 2")
+		return vmcommon.FunctionWrongSignature
+	}
+	token, returnCode := e.basicOwnerChecks(args)
+	if returnCode != vmcommon.Ok {
+		return returnCode
+	}
+	if !token.CanWipe {
+		e.eei.AddReturnMessage("cannot freeze")
+		return vmcommon.UserError
+	}
+	if len(args.Arguments[1]) != len(args.CallerAddr) {
+		e.eei.AddReturnMessage("invalid arguments")
+		return vmcommon.UserError
+	}
+
+	esdtTransferData := core.BuiltInFunctionESDTWipe + "@" + hex.EncodeToString(token.TokenName)
+	err := e.eei.Transfer(args.Arguments[1], e.eSDTSCAddress, big.NewInt(0), []byte(esdtTransferData), 0)
+	if err != nil {
+		e.eei.AddReturnMessage(err.Error())
+		return vmcommon.UserError
+	}
+
 	return vmcommon.Ok
 }
 
-func (e *esdt) wipe(_ *vmcommon.ContractCallInput) vmcommon.ReturnCode {
-	//TODO: implement me
+func (e *esdt) pauseUnPause(args *vmcommon.ContractCallInput, builtInFunc string) vmcommon.ReturnCode {
+	if len(args.Arguments) < 1 {
+		e.eei.AddReturnMessage("invalid number of arguments, wanted 2")
+		return vmcommon.FunctionWrongSignature
+	}
+	token, returnCode := e.basicOwnerChecks(args)
+	if returnCode != vmcommon.Ok {
+		return returnCode
+	}
+	if !token.CanPause {
+		e.eei.AddReturnMessage("cannot freeze")
+		return vmcommon.UserError
+	}
+	if len(args.Arguments[1]) != len(args.CallerAddr) {
+		e.eei.AddReturnMessage("invalid arguments")
+		return vmcommon.UserError
+	}
+	if token.IsPaused && builtInFunc == core.BuiltInFunctionESDTPause {
+		e.eei.AddReturnMessage("cannot pause an already paused contract")
+		return vmcommon.UserError
+	}
+	if !token.IsPaused && builtInFunc == core.BuiltInFunctionESDTUnPause {
+		e.eei.AddReturnMessage("cannot unPause an already paused contract")
+		return vmcommon.UserError
+	}
+
+	token.IsPaused = !token.IsPaused
+	err := e.saveToken(token)
+	if err != nil {
+		e.eei.AddReturnMessage(err.Error())
+		return vmcommon.UserError
+	}
+
+	esdtTransferData := builtInFunc + "@" + hex.EncodeToString(token.TokenName)
+	e.eei.TransferToAll(e.eSDTSCAddress, []byte(esdtTransferData))
+
 	return vmcommon.Ok
 }
 
-func (e *esdt) pause(_ *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+func (e *esdt) configChange(_ *vmcommon.ContractCallInput) vmcommon.ReturnCode {
 	//TODO: implement me
 	return vmcommon.Ok
 }
 
-func (e *esdt) unpause(_ *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+func (e *esdt) claim(_ *vmcommon.ContractCallInput) vmcommon.ReturnCode {
 	//TODO: implement me
 	return vmcommon.Ok
 }
 
-func (e *esdt) configChange(_ *vmcommon.ContractCallInput) vmcommon.ReturnCode {
-	//TODO: implement me
-	return vmcommon.Ok
+func (e *esdt) basicOwnerChecks(args *vmcommon.ContractCallInput) (*ESDTData, vmcommon.ReturnCode) {
+	if args.CallValue.Cmp(zero) != 0 {
```

Comment:
I don't think this check should be in `basicOwnershipChecks`. Or better so we don't have to figure out how to refactor, we can rename this method

Reply:
done.